### PR TITLE
No crash for unreachable commit

### PR DESCRIPTION
**Pull request by samueltardieu**
This is a partial fix to #814: the crash will no longer happen since the commit id unknown to jj will be considered non-existent.

# Checklist

If applicable:
- [X] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [X] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/revset.rs
Comment (martinvonz) [inline lib/src/revset.rs:94]: Since this code is not specific to Git, let's write it something like "Only recognize a commit if we've indexed it"

File: CHANGELOG.md
Comment (martinvonz) [inline CHANGELOG.md:56]: Nit: The bug is not specific to colocated repos. The reproduction I shared in the bug doesn't use a colocated repo, for example. So something like this might work:

```
Fixed a crash when trying to access a commit that's never been imported into the jj repo from a Git repo. They will now be considered as non-existent if referenced explicitly instead of crashing.
```

File: tests/test_git_colocated.rs
Comment (martinvonz) [inline tests/test_git_colocated.rs:252]: Optional: We don't care about the contents here, so you can simplify this by doing `let empty_tree_oid = git_repo.treebuilder(None).unwrap().write().unwrap();`. Then use that for both commits.

File: tests/test_git_colocated.rs
Comment (martinvonz) [inline tests/test_git_colocated.rs:292]: I think you pass `None` here so the branch doesn't get updated. That should make the reset below unnecessary.

File: tests/test_git_colocated.rs
Comment (martinvonz) [inline tests/test_git_colocated.rs:333]: nit: "cannot be found without crashing" makes it sound like we can find it if we crash :) Maybe something like "Check that trying to look up the second commit fails gracefully"

**Review by martinvonz — COMMENTED** [review]
Nice! Thanks for fixing that.

File: lib/src/revset.rs
Comment (samueltardieu) [reply lib/src/revset.rs]: Done

File: CHANGELOG.md
Comment (samueltardieu) [reply CHANGELOG.md]: Done.

File: tests/test_git_colocated.rs
Comment (samueltardieu) [reply tests/test_git_colocated.rs]: True. Done.

File: tests/test_git_colocated.rs
Comment (samueltardieu) [reply tests/test_git_colocated.rs]: Indeed. That's what I did first but I tried to be as close as possible to the reproducer mentionned in the issue. Done.

File: tests/test_git_colocated.rs
Comment (samueltardieu) [reply tests/test_git_colocated.rs]: Done.

File: lib/src/revset.rs
Comment (samueltardieu) [reply lib/src/revset.rs]: Done.

**Comment by samueltardieu** [comment]
Thanks @martinvonz for the thorough review, it gave me a chance to play with `jj` to apply those changes to the individual commits before squashing them.

**Comment by martinvonz** [comment]
Thanks again! Feel free to merge (I invited you as a collaborator).

**Comment by samueltardieu** [comment]
> Thanks again! Feel free to merge (I invited you as a collaborator).

It still requires an approval before it can be merged.

**Comment by martinvonz** [comment]
> > Thanks again! Feel free to merge (I invited you as a collaborator).
> 
> It still requires an approval before it can be merged.

Oops, I thought I did that.

**Review by martinvonz — APPROVED** [review]

**Pull request merged**
